### PR TITLE
week 4 assessment

### PR DESCRIPTION
**Pull request by yungRicardo**
# Submitting Week 4 Assessment

### ✅ Completed Files

Select the following as confirmation that the files in this repository are complete.

- [ ] Coding Challenges
- [ ] Coding Concepts
- [ ] Interview Questions

### ✋ Reminder

Do NOT merge your assessment!

### 📝 Feedback

If you would like specific feedback please leave details here:

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -27,10 +27,30 @@ const colors2 = [
...
+  return arr.sort((a, b) => Math.random() - 0.5)
+}
+// console.log(shuffleColors(colors1))
+// console.log(shuffleColors(colors2))
+
 // Pseudo code:
 
+// Create a function that takes in a array and eliminates the first element and shuffles the rest
+// Within function use .shift() to remove frist element
+// Return array with sort method to shuffle the rest, using Math.random()
```
Comment (chrisyhlee) [inline code-challenges.test.js:51]: Nice job, I'm a fan of the short but sweet pseudo code

File: code-challenges.test.js [hunk 2]
```diff
@@ -40,10 +60,28 @@ const votes1 = { upVotes: 13, downVotes: 2 }
...
+  return totalTally
+}
+
+
 // Pseudo code:
 
+// create function thats take obj returns end tally of upVotes and downVotes
+// inside code block declare new variable totalTally
+// assign object upVotes - downVotes to new variable
+// return totalTally
```
Comment (chrisyhlee) [inline code-challenges.test.js]: no need to make the variable, just return

File: code-challenges.test.js [hunk 3]
```diff
@@ -54,4 +92,25 @@ const dataTypesArray2 = ["string", "null", "Boolean", "string", "undefined"]
...
+mergedArr = [...dataTypesArray1, ...dataTypesArray2]
+console.log(dataTypes(...mergedArr))
+
+
 // Pseudo code:
+
+// Create function that accepts a dynamic number of arguments and returns an array with unique values
+// Declare variable and assign it the value [...new Set(data)]
+//the Set object accepts an array as arugment and returns a set with unique values. While the spread is used to hold the uniquw elements
+//we merge the the 2 arrays together using spread syntax and invoke dataTypes passing the mergedArr as a argument.
```
Comment (chrisyhlee) [inline code-challenges.test.js]: Nice attempt! Very close. I would pay attention to what you are receiving from `...data`. Make sure to console.log each step so you can see what each piece of code is returning and is what you're expecting.

File: interview-questions.md [hunk 4]
```diff
@@ -8,42 +8,44 @@ Then, research the question to expand on your answer. Even if you feel you have
...
-Your answer:
+Your answer: A block is chunk of code is that usually executed after a condtion is met or if a method invoked. 
 
-Researched answer:
+Researched answer: A block is a piece of code that can be passed to a method as an argument. 
 
 5. How do you extract a single value from a Ruby hash?
 
-Your answer:
+Your answer: to extract a value from a hash you must find the key that you trying to access.
```
Comment (chrisyhlee) [inline interview-questions.md]: I suggest giving more elaborate answers in all of your interview questions. Give an example of how to use it, and you can include how you personally have experience with the concept being asked about. The goal is for the interviewer to know that you know what you're talking about without a doubt.

File: ruby-challenges.rb [hunk 5]
```diff
@@ -32,3 +46,18 @@
 # Expected output: 'LEARN is not a palindrome'
 palindrome_test_case3 = 'Rotator'
 # Expected output: 'Rotator is a palindrome'
+
+def is_palindrome (str)
+    temp = str.downcase
+    if temp.reverse == temp
+        puts "#{str} is a palindrome"
+    elsif temp.reverse != temp 
+        puts "#{str} is not a palindrome"
```
Comment (chrisyhlee) [inline ruby-challenges.rb]: Make sure you don't leave any `puts` or `p` within your methods. It's essentially the same as leaving in console.logs in Javascript, which is bad practice.

**Review by chrisyhlee — CHANGES_REQUESTED** [review]
Missing a test for Javascript 3. This is probably the reason you didn't catch that your function is not working how you want it to as well. Please make the changes and resubmit

**Review by chrisyhlee — APPROVED** [review]
Thanks for making the requested changes. This assessment displays Advanced Beginner, but will be marked down to Novice because of tardiness. Take another look at your refactor for JS 3 and see why it's not working properly, it has to do with the spread operator in the parameter, which is probably not outputting what you think it is. You are approved to merge and delete the branch. I would focus on making your interview questions more elaborate by adding in examples and your personal experience with the concept being asked about. Otherwise good work!

**Comment by yungRicardo** [comment]
Could you please help me understand what I'm missing in JS challenge 3 please. I thought i fixed the problem

**Comment by chrisyhlee** [comment]
Sure, so if you console log what you have right now `console.log([...new Set(...data)] )`, you can see what it's returning, which is just the first array. So it's completely ignoring the second array entirely. To fix this, we need to combine the two arrays into one array, like we did during review. You can do this: `return [...new Set([].concat(...data))]`. This will take all the arrays passed into the function and put every value into one single array, and from there, we throw it in a Set to remove duplicates, and then use the spread operator in another array to put the Set values in an array. Hope this helped clarify some things, if there's still some confusion I'm happy to explain further.

**Pull request merged**